lib/recursive-methods.rb

```diff
@@ -1,49 +1,96 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), the amount of recursive calls depends on the size of n.
+# Space complexity: O(n), each recursive call is added to the stack. 
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 1 || n == 0
+  
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), the amount of recursive calls depends on the size of the string.
+# Space complexity: O(n), each recursive call is added to the stack. 
 def reverse(s)
```

Comment:
This works, but because you create a new array with each recursive call this is O(n<sup>2</sup>) for both time/space complexity.